### PR TITLE
[NFC] Refactor looping over recomputeLiveIns into function

### PR DESCRIPTION
**Pull request by redstar**
https://github.com/llvm/llvm-project/pull/79940 put calls to recomputeLiveIns into
a loop, to repeatedly call the function until the computation converges. However,
this repeats a lot of code. This changes moves the loop into a function to simplify
the handling.

Note that this changes the order in which recomputeLiveIns is called. For example,

```
  bool anyChange = false;
  do {
    anyChange = recomputeLiveIns(*ExitMBB) || recomputeLiveIns(*LoopMBB);
  } while (anyChange);
```

only begins to recompute the live-ins for LoopMBB after the computation for ExitMBB
has converged. With this change, all basic blocks have a recomputation of the live-ins
for each loop iteration. This can result in less or more calls, depending on the
situation.

### CODE REVIEW

**Comment by redstar** [comment]
My understanding is that the live-ins computation for the basic blocks is independent
of each other. If this is correct, then the loop could be formulated as follows, which avoids
unnecessary calls.

```
static inline void
fullyRecomputeLiveIns(ArrayRef<MachineBasicBlock *> MBBs) {
  for (MachineBasicBlock *MBB : MBBs)
    while (recomputeLiveIns(*MBB))
      /* intentionally empty/* ;
}
```

Also, is the function `recomputeLiveIns` in the current form needed? I see no callers to it, so both functions could be integrated into one.

**Comment by momchil-velikov** [comment]
> My understanding is that the live-ins computation for the basic blocks is independent of each other.
Live-ins are computed by backward data-flow analysis, thus live-ins of a block depend on the live-outs of the block, which 
in turn depend on the live-ins of the successor blocks.

> If this is correct, then the loop could be formulated as follows, which avoids unnecessary calls.
> ```
> static inline void
> fullyRecomputeLiveIns(ArrayRef<MachineBasicBlock *> MBBs) {
>   for (MachineBasicBlock *MBB : MBBs)
>     while (recomputeLiveIns(*MBB))
>       /* intentionally empty/* ;
> }
> ```

That's what I expect to result in unnecessary calls since once you've computed live-ins for a block, immediately (i.e. without touching a different block) re-computing them for the same block would result in no change (50% waste), except when the block loops into itself, in which case only you get "only" 33.3% waste.

> Also, is the function `recomputeLiveIns` in the current form needed? I see no callers to it, so both functions could be integrated into one.

I would suggest to leave it alone, so one can control independently the computation on a block and the order in which those computations are performed (even though DFS order is universally preferred).

**Review by momchil-velikov — APPROVED** [review]
LGTM. Perhaps it's worth checking CTMark times, I'd improvement (perhaps not measurable overall), but just to avoid surprises.

File: llvm/include/llvm/CodeGen/LivePhysRegs.h [hunk 1]
```diff
@@ -207,6 +209,18 @@ static inline bool recomputeLiveIns(MachineBasicBlock &MBB) {
   return oldLiveIns != newLiveIns;
 }
 
+/// Convenience function for recomputing live-in's for a set of MBBs until the
+/// computation converges.
+static inline void
```
Comment (yozhu) [inline llvm/include/llvm/CodeGen/LivePhysRegs.h:214]: This new function doesn't need to be defined as `static` in the header file.

**Comment by redstar** [comment]
> LGTM. Perhaps it's worth checking CTMark times, I'd improvement (perhaps not measurable overall), but just to avoid surprises.

Good point. I'll try a run.

**Review by oskarwirga — APPROVED** [review]
I tested these changes locally and they worked, LGTM!

**Comment by redstar** [comment]
> > LGTM. Perhaps it's worth checking CTMark times, I'd improvement (perhaps not measurable overall), but just to avoid surprises.
> 
> Good point. I'll try a run.

Sorry for the delay, I ran into some trouble with CTMark. Turned out that the commit I based my changes on had a problem.
I see a small degradation in compile time performance, which surprises me since I also checked some of the test cases, and the number of calls to `recomputeLiveIns()` decreased. I like to understand this first before committing the change.

File: llvm/include/llvm/CodeGen/LivePhysRegs.h (hunk 1 above)
Comment (redstar) [reply llvm/include/llvm/CodeGen/LivePhysRegs.h]: Fixed, thanks.

**Comment by redstar** [comment]
I removed all the syntactic sugar inside the loop. Now I get the expected result: with this change, clang is faster than the unmodified version (on CTMark).
Please have another look. Thanks.

**Review by yozhu — APPROVED** [review]

**Pull request merged**